### PR TITLE
Implement the get(name) function in Cookie Store API

### PR DESCRIPTION
**Pull request by RupinMittal**
#### 9ae242a9a2f095eed7c17d453874a733bd57ce92
<pre>
Implement the get(name) function in Cookie Store API
<a href="https://bugs.webkit.org/show_bug.cgi?id=259071">https://bugs.webkit.org/show_bug.cgi?id=259071</a>

Reviewed by Chris Dumez.

CookieStore (on the WebProcess side), will make an IPC to the
Network Process (via WebCookieJar) and will pass the relevant
CookieStoreGetOptions through the IPC. The Network process will
retrieve the stored cookies, filter them using the provided
options and will asynchronously resolve the promise with a
vector of the filtered cookies. In the get(name) function, the
promise is resolved with only the first cookie in the vector
(if there are any at all).

* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/cookie-store/CookieListItem.idl:
* Source/WebCore/Modules/cookie-store/CookieStore.cpp:
(WebCore::CookieStore::create):
(WebCore::CookieStore::CookieStore):
(WebCore::CookieStore::get):
(WebCore::CookieStore::activeDOMObjectName const):
(WebCore::CookieStore::scriptExecutionContext const):
* Source/WebCore/Modules/cookie-store/CookieStore.h:
* Source/WebCore/Modules/cookie-store/CookieStore.idl:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/loader/CookieJar.h:
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::cookieStore):
* Source/WebCore/platform/network/NetworkStorageSession.h:
* Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm:
(WebCore::NetworkStorageSession::vectorOfCookiesForSession const):
(WebCore::NetworkStorageSession::vectorOfCookiesForDOM const):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.cpp:
(WebKit::NetworkConnectionToWebProcess::cookiesForDOMAsync):
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.h:
* Source/WebKit/NetworkProcess/NetworkConnectionToWebProcess.messages.in:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp:
(WebKit::WebCookieJar::getCookiesAsync const):
* Source/WebKit/WebProcess/WebPage/WebCookieJar.h:

Canonical link: <a href="https://commits.webkit.org/265963@main">https://commits.webkit.org/265963@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/5e8e36d62db2d136a809c00e0ccb803a21ea0501

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/12420 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/12755 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/13066 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/14163 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/11939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/12479 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/15181 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/12773 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/14615 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/12584 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/13331 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/10499 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/14585 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/10614 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/11244 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/18343 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/11697 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/11409 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/14596 "Passed tests") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/11903 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9825 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/11136 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3050 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/15465 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/31/builds/11758 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | |

### CODE REVIEW

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (RupinMittal) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp:97]: I tried other ways of writing this line, but none of them seemed to work, so I kept this.

**Comment by cdumez** [comment]
The commit log is missing the reviewed by NOBODY OOPS line.

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: Seems odd to include Forward.h in a cpp.

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: Let's call this `options`:
`CookieStoreGetOptions options { name, { } };`

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: This should take a `Document&`, otherwise your downcast<Document> below looks unsafe.

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: Should take a Document& for now.

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp:95]: Most of this logic should be moved to the get() overload which takes in an Options struct. Then this function would just creates the Options struct and call the other get() function with it.

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: `promise->resolve<IDLDictionary<CookieListItem>>(CookieListItem { WTFMove(cookie.name), WTFMove(cookie.value) });`

File: Source/WebCore/Modules/cookie-store/CookieStore.h
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.h]: I doubt this is needed.

File: Source/WebCore/Modules/cookie-store/CookieStore.h
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.h]: explicit

File: Source/WebCore/loader/CookieJar.cpp
Comment (cdumez) [inline Source/WebCore/loader/CookieJar.cpp:193]: This should call the completion handler or it would crash in debug.

File: Source/WebCore/page/LocalDOMWindow.cpp
Comment (cdumez) [inline Source/WebCore/page/LocalDOMWindow.cpp]: `m_cookieStore = CookieStore::create(document());`

Nothing guarantees that `document()` returns non-null so the CookieStore Factory should take in a `Document*` not a `Document&`

File: Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm
Comment (cdumez) [inline Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm]: It is too wasteful to be doing this check *after* constructing the WebCore::Cookie object. You should do the check on the platform cookie type instead.

Comment (cdumez) [reply Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm]: Also, cookieStoreGetOptions.name can be null, in which case we want to return ALL cookies. The current check is wrong.

**Review by cdumez — CHANGES_REQUESTED** [review]

File: LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_arguments.https.any-expected.txt
Comment (RupinMittal) [inline LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_get_arguments.https.any-expected.txt:10]: I'm not sure that this change should have happened, but I put it here to get a second pair of eyes on it.

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: `get(CookieStoreGetOptions { WTFMove(name), { } }, WTFMove(promise));`

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: `String&&`

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: reject(Exception { SecurityError, "The origin is opaque"_s });

File: Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm
Comment (cdumez) [inline Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm]: Still wrong:
```
inline String::operator NSString *() const
{
    if (!m_impl)
        return @"";
    return *m_impl;
}
```

so if `options.name` is a null string, `name` will be an *empty* NSString (not `nil`) and you'll return no cookies instead of all of them.

File: Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp
Comment (cdumez) [inline Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp]: Missing period at the end of the comment.

File: Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp
Comment (cdumez) [inline Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp]: Missing period.

File: Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp
Comment (cdumez) [inline Source/WebCore/platform/network/curl/NetworkStorageSessionCurl.cpp]: missing `NetworkStorageSession::` ?

File: Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp
Comment (cdumez) [inline Source/WebCore/platform/network/soup/NetworkStorageSessionSoup.cpp]: missing `NetworkStorageSession::` ?

File: Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp
Comment (cdumez) [inline Source/WebKit/WebProcess/WebPage/WebCookieJar.cpp]: Why is this needed?

**Review by cdumez — CHANGES_REQUESTED** [review]

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: Could you please move this to the next line and call document.url() instead? It would avoid a virtual function call

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: The page needs null checking.

File: Source/WebCore/platform/network/NetworkStorageSession.h
Comment (cdumez) [inline Source/WebCore/platform/network/NetworkStorageSession.h]: Could we rename this to `cookiesForDOMAsVector()`. I think it would match better how we often name things in WebKit.

File: Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm
Comment (cdumez) [inline Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm]: I don't think this addressed my previous review comment? It still doesn't seem to deal with null Strings correctly.

File: Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm
Comment (cdumez) [inline Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm]: Similar concern about the naming, I'd use a `AsVector` suffix instead.

**Review by cdumez — CHANGES_REQUESTED** [review]

File: Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm
Comment (cdumez) [reply Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm]: Actually, I misread, you're null-checking `options.name` now so it should work. I'd recommend using `options.name.isNull()` for clarity.

File: Source/WebCore/Modules/cookie-store/CookieStore.cpp
Comment (cdumez) [inline Source/WebCore/Modules/cookie-store/CookieStore.cpp]: nit: Ideally we'd move this after the page null check too to avoid doing it when the page is null.

File: Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm
Comment (cdumez) [inline Source/WebCore/platform/network/cocoa/NetworkStorageSessionCocoa.mm]: nit:
```
if ([cookie isSecure] && includeSecureCookies == IncludeSecureCookies::No)
    continue;
```

would look nicer IMO.

**Review by cdumez — APPROVED** [review]
LGTM with nits, assuming the bots are happy.

**Pull request merged**
